src/views/DragAndDropController.ts

```diff
@@ -37,4 +60,154 @@ export class DragAndDropController implements TreeDragAndDropController<Explorer
+        // draggable node must have uri
+        if (!(node instanceof DataNode) || !node.uri) {
+            return;
+        }
+
+        // whether the node can be dropped will be check in handleDrop(...)
+        treeDataTransfer.set(Explorer.Mime.JavaProjectExplorer, new DataTransferItem(node.uri));
+    }
+
+    /**
+     * Handle the DnD event which comes from Java Project explorer itself.
+     * @param target the drop node.
+     * @param uri uri in the data transfer.
+     */
+    public async dropFromJavaProjectExplorer(target: ExplorerNode | undefined, uri: string): Promise<void> {
+        const source: DataNode | undefined = explorerNodeCache.getDataNode(Uri.parse(uri));
+        if (!this.isDraggableNode(source)) {
+            return;
+        }
+
+        if (!this.isDroppableNode(target)) {
+            return;
+        }
+
+        // check if the target node is source node itself or its parent.
+        if (target?.isItselfOrAncestorOf(source, 1 /*levelToCheck*/)) {
+            return;
+        }
+
+        if (target instanceof ContainerNode) {
+            if (target.getContainerType() !== ContainerType.ReferencedLibrary) {
+                return;
+            }
+
+            if (!(target.getParent() as ProjectNode).isUnmanagedFolder()) {
+                return;
+            }
+
+            // TODO: referenced library
+        } else if (target instanceof PackageRootNode || target instanceof PackageNode
+                || target instanceof FolderNode) {
+            await this.move(source!, target);
+        }
+    }
+
+    /**
+     * Check whether the dragged node is draggable.
+     * @param node the dragged node.
+     */
+    private isDraggableNode(node: DataNode | undefined): boolean {
+        if (!node?.uri) {
+            return false;
+        }
+        if (node instanceof WorkspaceNode || node instanceof ProjectNode
+                || node instanceof PackageRootNode || node instanceof ContainerNode
+                || node instanceof BaseSymbolNode) {
+            return false;
+        }
+
+       return this.isUnderSourceRoot(node);
+    }
+
+    /**
+     * Check whether the node is under source root.
+     *
+     * Note: There is one exception: The primary type directly under an unmanaged folder project,
+     * in that case, `true` is returned.
+     * @param node DataNode
+     */
+    private isUnderSourceRoot(node: DataNode): boolean {
+        let parent = node.getParent();
+        while (parent) {
+            if (parent instanceof ContainerNode) {
+                return false;
+            }
+
+            if (parent instanceof PackageRootNode) {
+                return parent.isSourceRoot();
+            }
+            parent = parent.getParent();
+        }
+        return true;
+    }
+
+    /**
+     * Check whether the node is able to be dropped.
+     */
+    private isDroppableNode(node: ExplorerNode | undefined): boolean {
```

Comment:
It's a little strange for me to drag a resource folder to a Java package: After I confirm, I can't find that folder in the destination package (since we will only show classes and packages under a package).

Reply:
I think that folder will become a child package of the dropped package. What's the behavior on your side?

Reply:
the package node shows green (new files exist) but the dragged files can't be found.

![dnd](https://user-images.githubusercontent.com/45906942/173327293-1caba87f-a802-48a7-9c10-c67a2f177058.gif)

Reply:
I found that in this case, Eclipse can show `static` as a normal folder, while we cannot.

The root cause is here: https://github.com/microsoft/vscode-java-dependency/blob/2a6c0fc9fa07480dd28d4f1f7214a946f9136f17/jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/PackageCommand.java#L406

We didn't take `IFolder` into consideration and just simply skip it.

I think we can still support moving resource folders, and fix the bug to show folders in packages.

Reply:
Filed an issue for this: https://github.com/microsoft/vscode-java-dependency/issues/635